Updated Key Management Scenarios (#38)
notaryproject/notaryproject: keymanagementrequirements.md

# Overview

> Key management for container signing can be broadly categorized into three general use cases:

Reviewer: > Key management for container signing 

Key management for artifact signing

- Key Setup/Signing (Key managment and integrations with client tooling for generating signatures.)
- Trust store configuration (Configuring which keys to trust for artifacts from a predefined source.)
- Signature Validity (Expiry, Revocation and/or Trusted Update List. Mechanism to update information on the validity of signatures.)

## Personas:

> - Publisher: User who builds and signs containers.

Reviewer: > Publisher: User who builds and signs containers.

 Publisher: User who builds and signs artifacts.

    - Publisher Admin: In some scenarios a publisher will include a group of users (i.e. teams or enterprises). Admin users (i.e. security administrator) will be responsible for configuring roots, provide access to use or generate delegate keys, and make decisions for key revocation.

>     - Publisher Builder: In some scenarios a publisher will include a group of users (i.e. teams or enterprises). Builders (i.e. developers, build hosts) will be responsible for creating artifacts. 

Reviewer: while we tend to think of enterprises in our cloud business, just suggesting we use less "big business" terms. 

This sentence works well without the (i.e. teams or enterprises)

+1 on the clarification of builders are developers and build hosts to indicated human and systems identities.

    - Publisher Signer: In some scenarios a publisher will include a group of users (i.e. teams or enterprises). Signers (i.e. developers, signing hosts) will be responsible for signing artifacts.

> - Deployer: User who deploys containers.

Reviewer: Same comment to remove (i.e. teams or enterprises) as the sentence works without it.
> Deployer: User who deploys containers.

Deployer: User who deploys artifacts

>     - Deployer Admin: In some scenarios a deployer will include a group of users (i.e. teams or enterprises). Admin users (i.e. security administrators) will be responsible for configuring deployment policies, including potentially a set of trusted roots or required signatures. 

Reviewer: ~~(i.e. teams or enterprises)~~

>     - Deployer Operator: In some scenarios a publisher will include a group of users (i.e. teams or enterprises). Operators (i.e. developers, deployment hosts) will be responsible for pulling container images from registries, verifying their signatures, and then running them.

Reviewer: ~~(i.e. teams or enterprises)~~

> be responsible for pulling container images from registries

be responsible for pulling artifacts images from registries

> - Repository Owner: In some scenarios a container may be stored in a repository that is managed by the repository owner. The repository may be public or private and could also be air-gapped.

Reviewer: There's a repo concept, and collection of repo concept where a collection of repos may live below a root repo.
This is akin to the docker hub "org" concept. 

`docker.io/org/repo`

In this case, a user may want to configure something at their org level, or the repo level.

I'm not sure we need to change anything here, but wanted to highlight this could represent a hierarchy where the repository may represent a single repo, or a collection of repos below the repo.

## Definitions:

> - Root key: A self signed key used for the lowest designation of trust. Root keys can be created by developers, organizations, public/private CAs, and registry operators. The root key should be retrieved from a trusted source that can establish the authenticity of the creator's identity.

Reviewer: Could _organizations_ be the community-friendly replacement for _enterprises_

- Signing key: A signing key is used to generate artifact signatures. A signing key should be signed with a root key or one of its intermediaries. The certificate chain with a signing key can be used to verify which root it belongs to. While a root key can be used as a signing key, this is not recommended as it creates a large blast radius and increases the risk of compromising a root key. 

> - Trust Store: The trust store defines the relationship between signing keys and artifacts that are used at validation time to determine whether to trust an artifact with a crytptographically valid signature. The trust store will relate a source (any source, specific registry, specific repository, or specific target) with a certificate (for root key, intermediate, or signing key) or key repository (for automated key distribtuion). It is not recommended to use a signing key as this will cause signature validation to fail if the signing key is rotated. 

Reviewer: > (any source, specific registry, specific repository, or specific target)

What does _target_ imply here?

- Trust Store Key Information: The key information configured in the trust store will be cryptographically verifiable and contain:
    - Public Key: Will be used to verify signed artifacts.
    - Signed Identity: Identity of creator of the signing key.
    - (Optional) Issuer: If the key is an intermediate this will describe the owner of the root.
    - Mechanism to update validity of signatures generated with the signing key.

## Requirements
- Signing an artifact MUST NOT require the publisher to perform additional actions with a registry/repository or registry operator. Uploading a signature MAY require additional actions.

> - Retrieving a signature MUST NOT require the deployer to perform additional actions with a registry/repository or registry operator beyond those required to pull an unsigned artifact.

Reviewer: Should we add a requirement for:

A user MUST be able to pull the signature and validate the content before pulling the content? 

This would mitigate trojan horse attacks where the content must be pulled and only later do they determine the content is problematic.

Reviewer 2: Pulling and validating the signature first then pull the content is beneficial in terms of engineering for saving network bandwidth.

However, in some scenarios where the client just want to dump / archive the content and signatures without any validation, they can pull in any order.

Reviewer: @shizhMSFT, is there something about the addition of...

> A user MUST be able to pull the signature and validate the content before pulling the content?

... that conflicts with the ability to pull the content and signatures? The note above is more related to assuring the signatures are separable from the content, so we can verify the signature before the content. It's not meant to imply a user can't pull content and signatures, just they must be separable.

- Artifact integrity, source, and signature expiry MUST be verifiable from the signature AND NOT require additional calls to the registry/repository.
- Signature allow list/deny list MAY require additional calls that will be defined in a separate document.

> - Moving an artifact from one repository to another SHOULD NOT invalidate the signature on the artifact.

Reviewer 3: Moving -> copying to avoid the implication that the original artifact is deleted in the process.

```suggestion
- Copying an artifact from one repository to another SHOULD NOT invalidate the signature on the artifact.
```

- Publishers SHOULD be able to sign with keys stored on their local machines, secure tokens, Hardware Security Modules (HSMs), or cloud based Key Management Services.
- Publishers SHOULD be able to generate multiple signatures for a single artifact.

> - Publisher admins MUST have a mechanism to revoke signatures to indicate they are no longer trusted. 

Reviewer: > Publisher admins MUST have a mechanism to revoke signatures to indicate they are no longer trusted. 

While I agree with the general sentiment on revocation, I believe we're saying revocation is an optional element. Can this be resolved with:

Publisher admins SHOULD have a mechanism to revoke signatures to indicate they are no longer trusted.

Reviewer 4: I disagree, I think they MUST have the mechanism, but they don't need to use it. But I'd be alright either way

Reviewer: I believe we're saying:
Users MUST have an OPTION to add key and signature revocation. It looks like we're culling down the initial scope to not require that feature be implemented, but users can.  And, we'll likely have some capabilities. 
So, if the feature isn't a MUST, just saying the requirement here can't be a MUST.
Is there another tweak we can make for clarity?

- Trust stores MUST be configurable by the deployer admin.

> - Deployer admins MUST be able to configure trusted entities for individual repositories and targets.

Reviewer: What does targets refer to here? Can we trim this back to repositories?

Reviewer 4: They also need to be able to configure who is trusted for individual artifacts. Maybe replace targets with artifacts?

Reviewer 3: Subdividing access within a repository doesn't follow the registry model. Push or pull access is granted to the entire repository. As a user I'm not sure there's a value add for Notary giving me the flexibility to trust a signer for one set of tags and not another within the repository.

Reviewer 4: Allowing the additional granularity would allow Notary to verify relationships beyond those supported by the registry's push access controls. I could see uses for this when a large team has repository push access, but there are a few more sensitive artifacts within the repository that should be signed by, say, a security team. But if no one would use this we can leave it at the repository level.

> - Deployers MUST be able to validate signatures on any version of an artifact including whether they have been revoked by the publisher.

Reviewer: Suggest SHOULD as we make revocation an optional element.

Reviewer 4: Revocation is optional for the publisher, but if there is a revocation made, the deployer MUST be able to see that it was revoked (they may choose to ignore this fact, but they should see it first).

> - Signature validation MUST be enforceable in air-gapped environments. 

Reviewer: +1

We might want to state an obvious here. 

> Signature validation MUST be enforceable in air-gapped environments. 

Signature validation MUST be enforceable in air-gapped environments without any external access. 

This accounts for the TSA goals, where the TSA can be verified in the air-gapped environment, as the TSA keys are capable of being copied to the Air-gapped environment.

> ## Requirements that need further discussion

Reviewer: Everything above looks solid, with minor nits in wording. 

In the spirit of getting agreed content merged, should content above be merged, with content below re-submitted as a separate PR or discussion?

### Signing Key Expiry

> In public key cryptography, a pair of private key a.k.a. signing key and public key a.k.a. verification key is generated, and later the public key is certified by a higher key in the hierarchy. The certification of the signing key has a validity period to designate how long it is valid for. This section will discuss tradeoff and recommendations for expiry times and signing key rotation.

Reviewer 3: "This section will discuss..." is this incomplete, or should the below sections be nested with a smaller heading level?

### External Timestamp Server
Time Stamping Authorities (TSAs) defined by RFC3161 provide signed timestamp for a signature in order to prove that the signature was generated during the validity period of a certificate.

#### Scenarios
With TSA, signature can be considered valid even if the signing cerificate is expired. This technique is widely used by Authenticode with SignTool, NuGet, Adobe Acrobat, and many other industrial products.

In the world of artifacts, including container images, scenarios are
- Developers sign their artifacts or images with certificates. The certificates MAY have a configurable expiry time.
- Developers publish their artifacts or images and at a later date stop maintaining them. Content consumers SHOULD be able to verify signatures until they expire and use the artifacts or images
- Attackers with compromised keys try to sign artifacts or images with timestamps before the key compromise event.

#### Pros and Cons

> There are many public TSA servers available on the Internet. The advantages of public timestamp servers are obvious:

Reviewer: ~~obvious~~

>     Public

Reviewer 2: Need to fix the markdown formatting.

Reviewer 2: @NiazFK It seems the links are also missing this in doc. Is it intended?

    Free

However, those public TSAs also come with disadvantages:
    Require Internet access for signing

>         It is not really a disadvantage since public TSAs are online services. Devices in the air-gapped environment SHALL access the Internet for timestamp signing services.

Reviewer: air-gapped environments MUST NOT require access "to the internet".

But, we can reword a bit differently:

> However, those public TSAs also come with disadvantages:
>    Require Internet access for signing

Content may be signed outside the air-gapped environment, and then transferred into the air-gap environment where it can be verified without public access.
Users requiring signing within an air-gapped environment MAY implement a TSA within their air-gapped environment.

Reviewer 2: Agreed.

Reviewer 4: For TSAs, I believe the internet is also required for verification to ensure that the signature was valid when created. There might be a way to pass this verification through an entity outside the air-gapped environment, but this does add an additional step.

    Out of the control of the signer

>         External dependency

Reviewer 4: This may be an issue for air gapped environments as they would need to access the TSA to determine whether a signature at a particular time should be trusted.

        Availability is not assured. No SLA on signing.

>             Not all TSAs are available in all regions.

Reviewer: Can we reword to:

Not all public TSAs are available in all global regions.

As we were discussing this, it seems TSAs may become services provided by vendors, filling the gaps in global access and reliability.

We may want to add verbiage that fallback TSAs can be used.

            Some regions may have high latency.
        The certificates of TSAs MAY be revoked at any time without notices.

>             The removal of trust of VeriSign broke .NET 5+ NuGet. Thus Microsoft has to disable the package verification with a new release to unblock customers.

Reviewer: I by no means want to hide how .NET got broken, as it's a great example for why we need fallback, or more reliable TSAs.
I do wonder if we should include these dated references within the specs, vs. discussions. 
Or, is this whole section a discussion?

Implications of not using a signed timestamp for a signature
    In the absense of additional timestamp signature, the signature is only considered valid till key expiry. This may limit the use of short lived keys.
    In case of key compromise it’s not possible to revoke signatures from a point in time where the time of compromise is known, as an attacker can create signatures with any signature time (by changing local time)

### Signature Expiry

Signatures can expire if

>     The certificate of the signing key or the timestamp key (whichever is longer) expires
>     The signed content indicates a expiry time

Reviewer 3: Removing timestamp key since that implies a specific implementation.

```suggestion
1. The certificate of the signing key expires
2. The signed content indicates a expiry time
```

Reviewer 3: Is this an either/or, both, or something we need to pick?

Thinking through the scenario of signing keys that quickly expire, and we say the signature expires when the signing key expires, then deployers need a way to see if the key can still be trusted, hopefully without updating each artifact signature.

Conversely if we say signing keys can create signatures that expire long after the signing key expires, I'd like to see a way to limit that, with a policy enforced on the publisher side. Perhaps that could happen with two expiration times in the signing key certificate (and parents in the chain), one for when they key is allowed to create a signature, and another for the max lifetime of the signatures it creates.

> In this section, the #2 scenario is focused.

Reviewer: > In this section, the #2 scenario is focused.

?

Reviewer 2: This whole section is ill-formated.

It should be

> 1. The certificate of the signing key or the timestamp key (whichever is longer) expires
> 2. The signed content indicates a expiry time
>
> In this section, the #2 scenario is focused.

An added benefit is to defend against freeze attack. Since the signature expires in a short time although the signing key can live longer, the verifiers have to obtain the latest signature to verify, which implies that the verifies have access to the latest content.

    Freeze Attack Similar to a replay attack, a freeze attack works by providing metadata that is not current. However, in a freeze attack, the attacker freezes the information a client sees at the current point in time to prevent the client from seeing updates, rather than providing the client older versions than the client has already seen. As with replay attacks, the attacker’s goal is ultimately to compromise a client who has vulnerable versions of packages installed. A freeze attack may be used to prevent updates in addition to having an installed package be out of date.

### Transparent Root key auto rotation

> Root keys form the basis of hierarchical trust systems, where intermediate and leaf keys chain back to a root key, and leaf keys are used to generate signatures. Consumers of signed artifacts associate a baseline level of trust with signatures that chain to roots they trust. In the scenario that a root key is no longer trusted, all signed artifacts chaining back to the root are no longer trusted. The conditions which render the root keys to be no longer trusted, are disclosed to the customer through some other channel (CVE, internal security audit, etc.). Safe rotation of keys require a overlap period where both new and old root keys are trusted. This is always not possible, such as when a root key is compromised.

Reviewer: > This is always **not** possible, such as when a root key is compromised.

This is not always possible, such as when a root key is compromised.

Reviewer 5: Ack

#### Scenarios

> 1. A public root associated with a key a Publisher uses to sign artifacts is compromised. The access to root key is compromised, rather than the root key itself being stolen, allowing an attacker to create new intermediate and leaf trusted keys.

Reviewer: > A public root associated with a key a Publisher uses to sign artifacts is compromised. The access to root key is compromised, rather than the root key itself being stolen, allowing an attacker to create new intermediate and leaf trusted keys.

A publisher uses a public root associated with a key to sign artifacts is compromised.

Reviewer 5: I meant to say it's the leaf certificate that's used to sign, not the root, I'll update the language to be clearer.

2. An internal root used by an organization is due to expire. A new root is created and rotated before the existing root expires.
3. An internal root used by an organization is not stored securely and the private key is lost. A new root is created to continue operations.

> 4. A cryptographic algorithm is deprecated, or a compliance standard requires a customer to upgrade their key strength. The customer wants to safely rotate a root key without rendering existing signatures invalid and disrupting operations.

Reviewer: ~~customer~~ - user

Reviewer 5: @SteveLasker could you clarify, is the suggestion to use *user* instead of *customer* in the entire doc? I plan to replace them with specific personas we defined in the next rev, *Publisher Admin* would be appropriate here.

Reviewer: @gokarmm, yeah. In the OSS community, I'm just trying to avoid _customer_ references as not all users are _customers_, but all _customers_ are _users_.

Using a persona name could help, but I imagine there are some cases where multiple personas would apply. So, just suggesting _user_ would be a bit more generic where multiple personas may apply

        - If signed allowlist are used, the artifacts themselves may not need to be signed.
    - Cons
        - The allowlist needs to be updated for every version of artifact being published
        - May need to maintain a large allowlist which may be an overhead to distribute
- Signature Denylist
    - For Consumers, denylist allows explicitly indicating that an artifact or dependency is untrusted
    - For Publishers, this allows communicating to Consumers that specific versions of artifact are untrusted.

- Centralized/local public/private lists
    - The list can be local to repository, requiring update to the list in each repository, and requires keeping track of all repositories where the artifact needs to be published/revoked.
    - The list can be centralized
        - Centralized deny list (e.g. CRL maintaned by public CAs) can be used. The endpoint infomation is included in the signature, and signature verification step checks against this list.
            - Customers can define network topology with restricted network access, where these endpoints may not be accesible for hosts where signature verification occurs.

>             - These endpoints may not be available in air-gapped environments.

Reviewer: While the endpoints would be unavailable, can we add content the lists can be replicated in the air-gapped environment? We discussed this a bit, where we can't replicate the dns with a different IP as these are not allowed in secure air-gapped environments. 
Just not sure how we allow alternative endpoints without creating man-in-the-middle attacks.

Reviewer 4: Any distribution mechanism would need to be secured to prevent a MitM attack (both for air-gapped environments and for everyone else). This can be done using a delegation from the existing trusted keys, or you could set up another equally trusted mechanism.

>     - Public lists (e.g. transparency logs) may not be suitable for enterprise customers who don’t want their artifact updates to be disclosed publicly.

Reviewer: ~~enterprise customers~~ - security concious users

Reviewer 4: maybe secrecy conscious users. There are different types of security, TLs are great for public auditing, not great for trade secrets.

> ## Prototype Stages

Reviewer: Great content. 
Should this be pulled out to an issue, to implement this as the next prototype?

# Overview
Key management for container signing can be broadly categorized into three general use cases:

> - Key Setup/Signing (Key managment and integrations with client tooling for generating signatures.)

Reviewer 4: ```suggestion
- Key Setup/Signing (Key management and integrations with client tooling for generating signatures.)
```

- Trust Store: The trust store defines the relationship between signing keys and artifacts that are used at validation time to determine whether to trust an artifact with a crytptographically valid signature. The trust store will relate a source (any source, specific registry, specific repository, or specific target) with a certificate (for root key, intermediate, or signing key) or key repository (for automated key distribtuion). It is not recommended to use a signing key as this will cause signature validation to fail if the signing key is rotated. 
- Trust Store Key Information: The key information configured in the trust store will be cryptographically verifiable and contain:
    - Public Key: Will be used to verify signed artifacts.
    - Signed Identity: Identity of creator of the signing key.
    - (Optional) Issuer: If the key is an intermediate this will describe the owner of the root.
    - (Optional) Mechanism to check whether signatures generated with the signing key have been rescinded.

> 

Reviewer 3: We should add a definition of "Policy". The capabilities of what a policy can provide would then be determined later in the doc.

Reviewer 4: Yes, it would also be good to define whether the policy is set by the publisher, deployer, or both (maybe a default by the publisher with the deployer able to overwrite). This will be helpful for thinking through the trust relationships.

- Attackers with compromised keys try to sign artifacts or images with timestamps before the key compromise event.

#### Pros and Cons

There are many public TSA servers available on the Internet. The advantages of public timestamp servers are obvious:
    Public
    Free

However, those public TSAs also come with disadvantages:
    Require Internet access for signing
        It is not really a disadvantage since public TSAs are online services. Devices in the air-gapped environment SHALL access the Internet for timestamp signing services.
    Out of the control of the signer
        External dependency
        Availability is not assured. No SLA on signing.
            Not all TSAs are available in all regions.
            Some regions may have high latency.
        The certificates of TSAs MAY be revoked at any time without notices.
            The removal of trust of VeriSign broke .NET 5+ NuGet. Thus Microsoft has to disable the package verification with a new release to unblock customers.

Implications of not using a signed timestamp for a signature

>     In the absense of additional timestamp signature, the signature is only considered valid till key expiry. This may limit the use of short lived keys.

Reviewer 4: This isn't always the case. If the system has a way to ensure consistency of information on the registry, then the delegating party can revoke the key before it's pre-set expiration time.

3. An root key used by a Publisher is close to expiry. A new key is created and rotated before the existing root expires.
4. An root key used by a Publisher is not stored securely and the private key is lost. A new key pair is created to continue operations.
5. A cryptographic algorithm is deprecated, or a compliance standard requires a Publisher to upgrade their key strength.

#### Approaches
1. Storage and distribution of trust policies from registry
Pros
- Trust policies need not be stored in another location
- Allows in band/automated distribution and update of trust policies to consumers.
Cons
- Automated update of trust policies can be disruptive to Deployers that consume artifacts. Deployers should be able to control when trust policies are updated.
- Only allows Publishers to define the trust policy. It works well when Publisher and Deployer are the same user/organization.
    - This model does not allow Deployers to independently define trust policies. Furthermore, it may be required that Deployer Admins define trusted publishers, and Deployer Operators only specify/configure which artifacts (from repositories) are deployed.

#### Recommendation

> - Trust policies SHOULD NOT be stored in the registry. Multiple Deployers can consume artifacts from the same repository, and may need to define trust policies independently.

Reviewer 4: I think that there is a middle ground here. Not every deployer will want to configure their own trust policies, so a default policy distributed by the registry lowers the bar for using Notary. However, as you say, deployers need a way to define their own policies, especially in the case of an untrusted registry. We can support both of these use cases.

- Trust policies SHOULD be configured, distributed/updated out of band from artifact updates from registry.

### Rescinding Signature Validity
Artifact Publishers need mechanisms to indicate that a signature they generated is still trustworthy. 

#### Scenarios

> - A Publisher vends a new version of an artifact, and wants to indicate the new version as trusted in addition to older versions already published.

Reviewer 4: A second scenario here is that they no longer want to certify a particular artifact (they mistakenly signed/discovered a vulnerability/etc).

- Trust Store: The trust store defines the relationship between signing keys and artifacts that are used at validation time to determine whether to trust an artifact with a crytptographically valid signature. The trust store will relate a source (any source, specific registry, specific repository, or specific target) with a certificate (for root key, intermediate, or signing key) or key repository (for automated key distribtuion). It is not recommended to use a signing key as this will cause signature validation to fail if the signing key is rotated. 
- Trust Store Key Information: The key information configured in the trust store will be cryptographically verifiable and contain:
    - Public Key: Will be used to verify signed artifacts.
    - Signed Identity: Identity of creator of the signing key.
    - (Optional) Issuer: If the key is an intermediate this will describe the owner of the root.
    - Mechanism to update validity of signatures generated with the signing key.

## Requirements
- Signing an artifact MUST NOT require the publisher to perform additional actions with a registry/repository or registry operator. Uploading a signature MAY require additional actions.

> - Retrieving a signature MUST NOT require the deployer to perform additional actions with a registry/repository or registry operator beyond those required to pull an unsigned artifact.

Reviewer 4: Well, the deployer will at least have to download the signature and perform any key management steps to see if the signature should be trusted.

- Artifact integrity, source, and signature expiry MUST be verifiable from the signature AND NOT require additional calls to the registry/repository.
- Signature allow list/deny list MAY require additional calls that will be defined in a separate document.
- Moving an artifact from one repository to another SHOULD NOT invalidate the signature on the artifact.
- Publishers SHOULD be able to sign with keys stored on their local machines, secure tokens, Hardware Security Modules (HSMs), or cloud based Key Management Services.
- Publishers SHOULD be able to generate multiple signatures for a single artifact.
- Publisher admins MUST have a mechanism to revoke signatures to indicate they are no longer trusted. 

> - Trust stores MUST be configurable by the deployer admin.

Reviewer 4: This requirement seems to imply an implementation that includes trust stores. In the interest of being implementation-neutral, maybe something like 'key management must be configurable by the deployer admin.

Reviewer: It's a good clarification. A vendor may choose to breakup the auth roles more granualrly.
